### PR TITLE
Feature/pedigree graph

### PR DESCRIPTION
**Pull request by airenzp**
This branch adds the pedigree graph to the individuals. The pedigree is built according to the number of generations specified, which by default is 5. One can navigate through the nodes in the tree using double click and use the navigation buttons for zoom and move. A herd pedigree can also be rendered accessing the herd view and url.
In order to avoid compilation issues, the volumes were removed from the frontend docker, since they are already added in the docker-compose file, and the paths updated.

### CODE REVIEW

File: .docker/frontend
Comment (danr) [inline .docker/frontend]: Why were these removed?

File: app/utils/data_access.py
Comment (danr) [inline app/utils/data_access.py]: Typo? `get_leaves`

File: frontend/genebank_pedigree.tsx
Comment (danr) [inline frontend/genebank_pedigree.tsx]: This is written in old-style class-based React which is not used elsewhere in the project.

File: frontend/genebank_pedigree.tsx
Comment (danr) [inline frontend/genebank_pedigree.tsx]: Prefer `const` (or `let`) over `var`

File: frontend/genebank_pedigree.tsx
Comment (danr) [inline frontend/genebank_pedigree.tsx]: There must be some way to change this without having to go through a full page reload?

File: frontend/individual_pedigree.tsx
Comment (danr) [inline frontend/individual_pedigree.tsx]: Again old-style class-based component

Comment (danr) [inline frontend/individual_pedigree.tsx]: This looks with a brief glance like an exact copy of the class from the other file... Perhaps some of the code can be shared?

File: frontend/package.json
Comment (danr) [inline frontend/package.json]: Why is this added? There are no flow types are there?

File: frontend/package.json
Comment (danr) [inline frontend/package.json]: Hmm I never needed to add this package explicitly to package.json

File: frontend/package.json
Comment (danr) [inline frontend/package.json]: Nor this

File: frontend/individual_pedigree.tsx
Comment (danr) [inline frontend/individual_pedigree.tsx]: This table is copied from some other file, isn't it?

**Review by danr — COMMENTED** [review]
Some comments

File: docker-compose.yml
Comment (pontus) [inline docker-compose.yml]: I think we still don't want to expose the database.

File: docker-compose.yml
Comment (pontus) [inline docker-compose.yml]: As discussed in slack, this will pollute the source directory, please revert/remove this from PR.

File: .docker/frontend
Comment (airenzp) [reply .docker/frontend]: one can create volumes  using docker-compose without declaring them in the Dockerfile. There was also an error with the paths, that referred to a frontend folder that did not exist in the dockerfile but did in the docker-compose. I commented this before with Pontus in Slack...Hope this makes it clearer.

File: app/utils/data_access.py
Comment (airenzp) [reply app/utils/data_access.py]: yes, I will fix it, sorry.

File: frontend/genebank_pedigree.tsx
Comment (airenzp) [reply frontend/genebank_pedigree.tsx]: Mmmm, ok. This is my first react app. I will check it.

File: frontend/genebank_pedigree.tsx
Comment (airenzp) [reply frontend/genebank_pedigree.tsx]: ok

File: frontend/genebank_pedigree.tsx
Comment (airenzp) [reply frontend/genebank_pedigree.tsx]: ok, will check it

File: frontend/individual_pedigree.tsx
Comment (airenzp) [reply frontend/individual_pedigree.tsx]: ok

Comment (airenzp) [reply frontend/individual_pedigree.tsx]: ok

File: frontend/package.json
Comment (airenzp) [reply frontend/package.json]: probably can be removed. I was trying to solve my compilation issues at first installing what it seemed missing libraries, according to the error. I will check it.

File: frontend/package.json
Comment (airenzp) [reply frontend/package.json]: same here

File: frontend/package.json
Comment (airenzp) [reply frontend/package.json]: same here, as in my previous comment, will check it

File: frontend/individual_pedigree.tsx
Comment (airenzp) [reply frontend/individual_pedigree.tsx]: what do yo mean?

File: frontend/genebank_pedigree.tsx
Comment (norling) [reply frontend/genebank_pedigree.tsx]: yeah, try letting react do the routing with `react-router-dom`. You can import `useHistory` and then do `history.push` to navigate. This can be useful if you wish to leave a "trail" so that people can go back through their changes by navigating back. If you just want to update you can use the functions in the `communication.tsx` file to fetch data asynchronously!

File: frontend/individual_pedigree.tsx
Comment (norling) [inline frontend/individual_pedigree.tsx]: Since we're using material-ui for the styles in most of the app - consider using their tables ([here](https://material-ui.com/components/tables/)) to get a consistent look on the page easily! :)

File: app/herdbook.py
Comment (norling) [inline app/herdbook.py:28]: I don't mind using the Flask login system, but I'm not a big fan of using different systems on different endpoints. I seem to remember that the flask login system didn't really play that well with asynchronous requests, but if you can get it to work - please change all the endpoints so that the app remains consistent! :)

**Review by norling — CHANGES_REQUESTED** [review]
Some comments that are mostly up to you, but I would really like if you'd look over the login system before we merge!

File: app/herdbook.py
Comment (airenzp) [reply app/herdbook.py:28]: I did not touch your login system, just added what I needed to use the login_required attribute on the python methods. This does not change the react form or any logic behind. But I would like to be able to implement the return to an url after login, for example. For that, some integration with react needs to be done, since the return url is sent to the login page

File: frontend/individual_pedigree.tsx
Comment (airenzp) [reply frontend/individual_pedigree.tsx]: ok, thanks. I will take a look at that.

File: frontend/genebank_pedigree.tsx
Comment (airenzp) [reply frontend/genebank_pedigree.tsx]: thanks, ok

File: app/herdbook.py
Comment (norling) [reply app/herdbook.py:28]: Absolutely! But if we use decorators for one endpoint we should use them for all endpoints that require login so that the system remains consistent. Adding a login-redirector sounds like a great plan though! :)

Comment (airenzp) [reply app/herdbook.py:28]: Sure. I just did not wanted to do that without your approval

File: frontend/individual_pedigree.tsx
Comment (airenzp) [reply frontend/individual_pedigree.tsx]: I used the individual.tsx file as a starting point and added the pedigree graph.

File: app/utils/data_access.py
Comment (airenzp) [reply app/utils/data_access.py]: solved

File: frontend/genebank_pedigree.tsx
Comment (airenzp) [reply frontend/genebank_pedigree.tsx]: solved

File: frontend/genebank_pedigree.tsx
Comment (airenzp) [reply frontend/genebank_pedigree.tsx]: used history

File: frontend/package.json
Comment (airenzp) [reply frontend/package.json]: solved

File: frontend/package.json
Comment (airenzp) [reply frontend/package.json]: solved

File: frontend/package.json
Comment (airenzp) [reply frontend/package.json]: solved

File: frontend/genebank_pedigree.tsx
Comment (airenzp) [reply frontend/genebank_pedigree.tsx]: solved

File: frontend/individual_pedigree.tsx
Comment (airenzp) [reply frontend/individual_pedigree.tsx]: solved

Comment (airenzp) [reply frontend/individual_pedigree.tsx]: solved

File: frontend/individual_pedigree.tsx
Comment (airenzp) [reply frontend/individual_pedigree.tsx]: should I remove the previous individual.tsx?

File: frontend/individual_pedigree.tsx
Comment (airenzp) [reply frontend/individual_pedigree.tsx]: solved

File: app/herdbook.py
Comment (airenzp) [reply app/herdbook.py:28]: done

**Review by pontus — CHANGES_REQUESTED** [review]
Nice work

File: docker-compose.yml
Comment (airenzp) [reply docker-compose.yml]: ok

File: docker-compose.yml
Comment (airenzp) [reply docker-compose.yml]: As I said before, these volumes are already in the develop branch, they are not my creation. I just fixed the path. If we do not want them, then Dan needs to approve this, as it is him who added them. I do not know the effects of removing the volumes.

Comment (airenzp) [reply docker-compose.yml]: @danr you might want to comment your ideas on this matter.

File: frontend/individual_pedigree.tsx
Comment (danr) [reply frontend/individual_pedigree.tsx]: I don't know, up to you and Martin :)

File: frontend/genebank_pedigree.tsx
Comment (danr) [reply frontend/genebank_pedigree.tsx]: Nice! :+1:

File: frontend/genebank_pedigree.tsx
Comment (danr) [reply frontend/genebank_pedigree.tsx]: Perfect! :+1:

File: docker-compose.yml
Comment (danr) [reply docker-compose.yml]: I don't know, I'm very confused about the docker volumes for the cache directories. If you're happy with this setup and don't run into caching problems, go for it!

File: frontend/pedigree.tsx
Comment (danr) [inline frontend/pedigree.tsx]: Note that refs from useRef don't ever change so these don't need to be listed in the dependency list, see https://reactjs.org/docs/hooks-reference.html#useref
I think the reason why this still works is because the domNode ref is updated before your callback is run. One alternative would be to put the ref in a state instead, as in https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node

You don't need to list options in the dependency list either since it is a constant reference (declared with `const`!)

File: frontend/pedigree.tsx
Comment (danr) [inline frontend/pedigree.tsx]: I don't know how much "typing coverage" you're aiming for in this project but adding types here can be done like so:

```typescript
const PedigreeNetwork = ({pedigree}: {pedigree: string}) => {...
```

Also if you define components with `function` you see that name in debugging views (profiler, componentsv iew and error messages).

File: frontend/pedigree.tsx
Comment (danr) [inline frontend/pedigree.tsx]: I don't usually use default exports but if you want to use it you can export it default in one go

```typescript
export default function PedigreeNetwork(...) { 
    ...
}
```

**Review by danr — APPROVED** [review]

File: docker-compose.yml
Comment (pontus) [reply docker-compose.yml]: As I said in slack, what you do here is essentially removing the use of modules - they are still mounted under /frontend/node_modules etc. Whereas things are now run under /app (which will lead to the source directory being polluted by the cache folders and so on).

If you want to change to `/app`, I'd really want the volumes to be mounted there as well - but I suspect that will bring back your issue with the volume being corrupted over time (so I suspect you may want to run with the current setup in your specific environment).

File: frontend/pedigree.tsx
Comment (airenzp) [reply frontend/pedigree.tsx]: Ok, all these concepts are still quite loose for me. But I am looking forward to properly study react! Nice to have someone to point in the right direction.

File: docker-compose.yml
Comment (airenzp) [reply docker-compose.yml]: That might work, what was causing my compilation issues was having different volumen configurations in the dockerfile and the docker-compose. So you want to put everything inside the app folder instead? I am testing it now.

Comment (pontus) [reply docker-compose.yml]: I'm fairly convinced that what you ran into was a bug for Docker on macOS (which quite possibly can come back if you start using the volumes again). I don't much care about the actual path used inside the container, but I do not want temporary files and node libraries to spill outside of the container.

Comment (airenzp) [reply docker-compose.yml]: In the original code, there was this line in the dockerfile:
VOLUME /app/node_modules

And then in the docker-compose:
volumes:
            - type: bind
              source: ./frontend
              target: /frontend
              - /frontend/node_modules

You are saying in one place that your volume in the container is under the /app folder and in the other that the same volume is under the /frontend folder. Do we agree on this?

File: frontend/pedigree.tsx
Comment (danr) [reply frontend/pedigree.tsx]: Seems like you're picking it up very quickly! Feel free to ask questions in the slack, there's the #frontend channel which has mainly been used to discuss React pecularities so far :)

File: docker-compose.yml
Comment (pontus) [reply docker-compose.yml]: These would be independent (although possibly at some time in history intended to be the same, and that doesn't hurt but is not a strict requirement either).

The benefit of having the same path in both places is that the same volume is used (without going through to much trouble) and thus that the volume is "preloaded" on the first run of the instance. This means the `yarn install` and so on runs faster on the first launch.

Do note that for what happens in the container image, it's not unimportant whatever the `VOLUME` is used or not, having it will move what's under the given path to the new volume and thus it will not be included in the container image (it doesn't

That is not the same as the issue of changing some paths where things are mounted in the container but not other (e.g. changing the bind mount from `/frontend` to `/app` while still having the volumes at `/frontend/node_modules` and so on).

The `VOLUME` keyword in the Dockerfile will also register the volume for the container, so in theory, things should "just work" if one continue to use the same path. In practice, I've found that behaviour inconsistent, at least when having other mountpoints possibly interfering with it, so I prefer setting them up explicitly (which has been consistent in my experience).

Comment (airenzp) [reply docker-compose.yml]: It seems like we are making some progress here.
Then, what do you want me to do?

Comment (pontus) [reply docker-compose.yml]: Anything that won't spill the contents that did go in the volume over to the bind mounted source directory is fine, this could be:

* reverting this change
* keeping the change but also changing the volumes so they are mounted as subdirectories in the same directory as the source  is mounted

Comment (airenzp) [reply docker-compose.yml]: like this?
volumes:
            - type: bind
              source: ./frontend
              target: /app
            - ./frontend/node_modules:/app/node_modules
            - ./frontend/.parcel-cache:/app/.parcel-cache
            - ./frontend/.cache:/app/.cache

Comment (pontus) [reply docker-compose.yml]: IIRC that will be interpreted as additional bind mounts, which is what I don't want.

```
volumes:
  - type: bind
    source: ./frontend
    target: /app
  - /app/node_modules
  - /app/.parcel-cache
  - /app/.cache
```

should be fine.

Comment (airenzp) [reply docker-compose.yml]: Great, it works. I have now pushed this version

**Review by pontus — APPROVED** [review]

File: frontend/pedigree.tsx
Comment (danr) [inline frontend/pedigree.tsx]: You should still be able to remove the Refs from the dependency list (domNode and network)

Comment (airenzp) [reply frontend/pedigree.tsx]: I just did it. Thanks

File: frontend/pedigree.tsx
Comment (airenzp) [reply frontend/pedigree.tsx]: done

File: frontend/pedigree.tsx
Comment (airenzp) [reply frontend/pedigree.tsx]: How can I see the frontend channel? I need to be added maybe?

File: frontend/individual_pedigree.tsx
Comment (danr) [inline frontend/individual_pedigree.tsx]: I would add `id` to the depedency list since there is a dependency.

File: frontend/individual_pedigree.tsx
Comment (danr) [inline frontend/individual_pedigree.tsx:18]: Note that generations_input won't change if the generations prop (input) is changed (from parent).

One way to sync them up is to add an effect depending on `generations` that sets this state to `generations` when it changes.

Comment (airenzp) [reply frontend/individual_pedigree.tsx:18]: But the generations prop is not expected to change, this value is read from the url parameters, like the id in the previous comment. I thought this dependency was meant to be used for a changing parameter.

File: frontend/individual_pedigree.tsx
Comment (airenzp) [reply frontend/individual_pedigree.tsx]: ok

File: frontend/individual_pedigree.tsx
Comment (airenzp) [reply frontend/individual_pedigree.tsx:18]: ok

Comment (danr) [reply frontend/individual_pedigree.tsx:18]: Right one can have props that are not supposed to change. One way to enforce that they don't change is to remount the component using `key`: 
https://www.saltycrane.com/blog/2019/12/how-remount-react-component-when-prop-changes/
I think this is slightly weird since this is determined at the call site of the component rather than at the definition site of the component. It would make sense if one could indicate this when defining the component.

File: frontend/individual_pedigree.tsx
Comment (norling) [inline frontend/individual_pedigree.tsx]: this comment is rendered on the page :) it needs to be wrapped up as `{/* comment */}`

**Review by norling — APPROVED** [review]

File: frontend/individual_pedigree.tsx
Comment (airenzp) [reply frontend/individual_pedigree.tsx]: upss

Comment (airenzp) [reply frontend/individual_pedigree.tsx]: solved

**Pull request merged**
